wakuv2/missing_messages.go

```diff
@@ -0,0 +1,257 @@
+package wakuv2
+
+import (
+	"context"
+	"encoding/hex"
+	"errors"
+	"slices"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/peer"
+	"google.golang.org/protobuf/proto"
+
+	"go.uber.org/zap"
+
+	gethcommon "github.com/ethereum/go-ethereum/common"
+
+	"github.com/waku-org/go-waku/waku/v2/protocol"
+	"github.com/waku-org/go-waku/waku/v2/protocol/store"
+
+	"github.com/status-im/status-go/logutils"
+	"github.com/status-im/status-go/wakuv2/common"
+
+	"github.com/waku-org/go-waku/waku/v2/protocol/pb"
+)
+
+const maxContentTopicsPerRequest = 10
+const maxAttemptsToRetrieveHistory = 3
+const delay = 10 * time.Second
+
+type TopicInterestPerStorenode struct {
+	peerID        peer.ID
+	pubsubTopic   string
+	contentTopics []string
+	lastChecked   time.Time
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func (p TopicInterestPerStorenode) Equals(other TopicInterestPerStorenode) bool {
+	if p.peerID != other.peerID {
+		return false
+	}
+
+	if p.pubsubTopic != other.pubsubTopic {
+		return false
+	}
+
+	slices.Sort(p.contentTopics)
+	slices.Sort(other.contentTopics)
+
+	for i, contentTopic := range p.contentTopics {
+		if contentTopic != other.contentTopics[i] {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (w *Waku) CheckForMissingMessages(peerID peer.ID, pubsubTopic string, contentTopics []string) error {
+	w.topicInterestMu.Lock()
+	defer w.topicInterestMu.Unlock()
+
+	ctx, cancel := context.WithCancel(w.ctx)
+	newMissingMessageRequest := TopicInterestPerStorenode{
+		peerID:        peerID,
+		pubsubTopic:   pubsubTopic,
+		contentTopics: contentTopics,
+		lastChecked:   w.timesource.Now().Add(delay),
+		ctx:           ctx,
+		cancel:        cancel,
+	}
+
+	currMessageVerificationRequest, ok := w.topicInterest[pubsubTopic]
+	if !ok || !currMessageVerificationRequest.Equals(newMissingMessageRequest) {
+		if ok {
+			currMessageVerificationRequest.cancel()
+		}
+
+		w.topicInterest[pubsubTopic] = newMissingMessageRequest
```

Comment:
Hm, I think the problem here is that this function shouldn't return an error at all. I'll do the change and add a comment indicating the intention of this code. What this is supposed to do is in case there's a change in the list of content topics to verify, to cancel any active request, and replace the content topics with a new list.